File: extensions/code-exec/runner.py
"""Persistent Python execution kernel for the Pi code-exec extension.

Reads NDJSON commands from stdin, executes code in a persistent namespace, and
writes one NDJSON result line per command to the original stdout. Captured
print() output plus the value of a trailing expression are returned each time;
variables and imports persist across commands until a reset.
"""

import ast
import contextlib
import io
import json
import sys
import traceback

_globals = {"__name__": "__main__"}
# 真实 stdout：执行期间 sys.stdout 会被重定向去捕获 print，结果走这个原始句柄。
_out = sys.stdout


def _run(code):
    buf_out, buf_err = io.StringIO(), io.StringIO()
    value = None
    ok = True
    error = None
    try:
        tree = ast.parse(code, mode="exec")
        last_expr = None
        # 末尾若是表达式，单独 eval 取值（像 REPL 回显），其余语句正常 exec。
        if tree.body and isinstance(tree.body[-1], ast.Expr):
            last_expr = ast.Expression(tree.body.pop().value)
        with contextlib.redirect_stdout(buf_out), contextlib.redirect_stderr(buf_err):
            if tree.body:
                exec(compile(tree, "<cell>", "exec"), _globals)
            if last_expr is not None:
                result = eval(compile(last_expr, "<cell>", "eval"), _globals)
                if result is not None:
                    value = repr(result)
    except BaseException:  # 含 KeyboardInterrupt：捕获后内核存活，仅本条 exec 失败
        ok = False
        error = traceback.format_exc()
    return {
        "stdout": buf_out.getvalue(),
        "stderr": buf_err.getvalue(),
        "value": value,
        "ok": ok,
        "error": error,
    }


def _emit(obj):
    _out.write(json.dumps(obj, ensure_ascii=False) + "\n")
    _out.flush()


def main():
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            msg = json.loads(line)
        except Exception:
            continue
        mtype = msg.get("type")
        mid = msg.get("id")
        if mtype == "exec":
            res = _run(msg.get("code", ""))
            res["type"] = "result"
            res["id"] = mid
            _emit(res)
        elif mtype == "reset":
            _globals.clear()
            _globals["__name__"] = "__main__"
            _emit({"type": "result", "id": mid, "stdout": "", "stderr": "",
                   "value": None, "ok": True, "error": None})
        elif mtype == "ping":
            _emit({"type": "pong", "id": mid})
        else:
            _emit({"type": "result", "id": mid, "stdout": "", "stderr": "",
                   "value": None, "ok": False, "error": "unknown command type: %r" % (mtype,)})


if __name__ == "__main__":
    main()

File: extensions/code-exec/runner.mjs
// 常驻 JS 执行内核（Pi code-exec 的 js_run/js_reset）。
// 读 stdin NDJSON，在 node:vm 持久 context 执行 JS，回写一行 NDJSON 结果。
// completion value 回显（vm 返回末尾表达式值）；var/全局赋值跨 cell 持久（顶层 let/const 块级不持久）。
import vm from "node:vm";
import { inspect } from "node:util";

function makeContext() {
  const sandbox = {};
  sandbox.globalThis = sandbox;
  return vm.createContext(sandbox);
}

let context = makeContext();

function fmt(args) {
  return args.map((a) => (typeof a === "string" ? a : inspect(a, { depth: 4 }))).join(" ");
}

function run(code) {
  let stdout = "";
  let stderr = "";
  context.console = {
    log: (...a) => {
      stdout += `${fmt(a)}\n`;
    },
    info: (...a) => {
      stdout += `${fmt(a)}\n`;
    },
    debug: (...a) => {
      stdout += `${fmt(a)}\n`;
    },
    warn: (...a) => {
      stderr += `${fmt(a)}\n`;
    },
    error: (...a) => {
      stderr += `${fmt(a)}\n`;
    },
  };
  let value = null;
  let ok = true;
  let error = null;
  try {
    const result = vm.runInContext(code, context, { filename: "<cell>" });
    if (result !== undefined) value = inspect(result, { depth: 4 });
  } catch (e) {
    ok = false;
    error = e && e.stack ? String(e.stack) : String(e);
  }
  return { stdout, stderr, value, ok, error };
}

function emit(obj) {
  process.stdout.write(`${JSON.stringify(obj)}\n`);
}

function handle(line) {
  const s = line.trim();
  if (!s) return;
  let msg;
  try {
    msg = JSON.parse(s);
  } catch {
    return;
  }
  const id = msg.id;
  if (msg.type === "exec") {
    emit({ type: "result", id, ...run(msg.code ?? "") });
  } else if (msg.type === "reset") {
    context = makeContext();
    emit({ type: "result", id, stdout: "", stderr: "", value: null, ok: true, error: null });
  } else if (msg.type === "ping") {
    emit({ type: "pong", id });
  } else {
    emit({ type: "result", id, stdout: "", stderr: "", value: null, ok: false, error: `unknown command type: ${msg.type}` });
  }
}

let buf = "";
process.stdin.setEncoding("utf8");
process.stdin.on("data", (chunk) => {
  buf += chunk;
  let idx = buf.indexOf("\n");
  while (idx >= 0) {
    handle(buf.slice(0, idx));
    buf = buf.slice(idx + 1);
    idx = buf.indexOf("\n");
  }
});
